kvdb-web/src/indexed_db.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility functions to interact with IndexedDB browser API.
+
+use wasm_bindgen::{JsCast, JsValue, UnwrapThrowExt, closure::Closure};
+use web_sys::{
+	IdbDatabase, IdbRequest, IdbOpenDbRequest,
+	Event, IdbCursorWithValue,
+	IdbTransactionMode,
+};
+use js_sys::{Array, Uint8Array, ArrayBuffer};
+
+use futures::channel;
+use futures::prelude::*;
+
+use kvdb::{DBOp, DBTransaction};
+
+use std::ops::Deref;
+use log::{debug, warn};
+
+
+use crate::Column;
+
+
+/// Opens the IndexedDB with the given name and the specified number of columns
+/// (including the default one).
+pub fn open(name: &str, columns: u32) -> impl Future<Output = IdbDatabase> {
+	let (tx, rx) = channel::oneshot::channel::<IdbDatabase>();
+	// TODO: handle errors more gracefully,
+	// return a Result instead of expect_throw?
+	let window = web_sys::window().expect_throw("are we in a browser?");
+	let indexed_db = window.indexed_db()
+		.expect_throw("IndexDB should be supported in your browser")
+		.expect_throw("IndexDB should be supported in your browser");
+
+	// TODO: don't hardcode version
+	let open_request = indexed_db.open_with_u32(name, /* version */ 1)
```

Comment:
There are only two ways to handle this "version" thing here IMO: either we use the version, by passing it as parameter and adding a method to retrieve it later, or we don't pass any version at all.
I would go for no version at all.
```suggestion
	let open_request = indexed_db.open_with_u32(name)
```

Reply:
Currently, if we reopen the database with more columns that it was opened previously, it will panic on open, when trying to create a transaction to read from a new column. Because a new column is expected to be created on version upgrade (`try_create_object_stores`).

We can pass the number of columns as a version, sounds reasonable?

Reply:
Isn't it possible to dynamically check the number of columns, and add the missing ones?

Reply:
I've tried that, but creating a column only possible onupgradeneeded:
> The createObjectStore() method of the IDBDatabase interface creates and returns a new object store or index.
> this method can be called only within a versionchange transaction.

Reply:
I see! Then bump the version if we detect that the number of columns is different?
It might seem stupid, but I'd prefer to use things as they are intended to be used. In other words, the "version" should actually be a version and not a way for us to store a number.

Reply:
> Then bump the version if we detect that the number of columns is different?

The version is bumped only when we open the db with a higher version. And we can query the version on after either the open requests succeeds or `onupgradeneeded` is called (we specified a higher version that it is). 

So I went with "by passing it as parameter and adding a method to retrieve it later" approach.

Reply:
> So I went with "by passing it as parameter and adding a method to retrieve it later" approach.

What worries me is that nobody is probably ever going to use this mechanism.
RocksDB automatically updates the number of columns if it doesn't match the number on disk, and I'm not sure it's a good idea to force users do that manually for IndexedDB.

Reply:
In parity-ethereum we have a separate version file stored for RocksDB, see https://github.com/paritytech/parity-ethereum/blob/master/parity/db/rocksdb/migration.rs and the it refuses to open the newer version. 

We could open the latest version of the db, see the version and the number of columns, and if the we need more columns, close it and reopen with a different number of version. But that sounds like a hack.

Reply:
FWIW I've implemented this hacky solution.